branch.go

```diff
@@ -0,0 +1,196 @@
+package flow
+
+import (
+	"context"
+	"sync"
+)
+
+// BranchCheckFunc checks the target and returns true if the branch should be taken.
+type BranchCheckFunc[T Steper] func(context.Context, T) (bool, error)
+
+// If adds a conditional branch to the workflow.
+//
+//	If(someStep, func(ctx context.Context, someStep *SomeStep) (bool, error) {
+//		// branch condition here, true -> Then, false -> Else.
+//		// if error is returned, then fail the selected branch step.
+//	}).
+//	Then(thenStep).
+//	Else(elseStep)
+func If[T Steper](target T, check BranchCheckFunc[T]) *IfBranch[T] {
+	return &IfBranch[T]{Target: target, BranchCheck: BranchCheck[T]{Check: check}}
+}
+
+// IfBranch adds target step, then and else step to workflow,
+// and check the target step and determine which branch to go.
+type IfBranch[T Steper] struct {
+	Target      T // the target to check
+	BranchCheck BranchCheck[T]
+	ThenStep    []Steper
+	ElseStep    []Steper
+	Cond        Condition // Cond is the When condition for both ThenStep and ElseStep, not target Step!
+}
+
+// Then adds steps to the Then branch.
+func (i *IfBranch[T]) Then(th ...Steper) *IfBranch[T] {
+	i.ThenStep = append(i.ThenStep, th...)
+	return i
+}
+
+// Else adds steps to the Else branch.
+func (i *IfBranch[T]) Else(el ...Steper) *IfBranch[T] {
```

Comment:
Not sure if we can restrict that `flow.If().Then().Else()` is the only allowed format. If we write `flow.If().Else().Then()`, that's weird.

Reply:
Yes. How to prevent this weird usage? Or allows it?

Reply:
We cannot as long as we use this builder pattern....
> “You can't wake a person who is pretending to be asleep”